### PR TITLE
Keycard installation

### PR DESCRIPTION
**Pull request by dmitryn**
### Summary:

Implements keycard installation. Screens are based on https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Hardwallet-light-flows?node-id=0%3A1

### Review notes

Sorry for big PR, but at this point it's hard to break this feature (keycard installation process) into smaller parts. Although, after this PR merged, new features (like installation progress bar, login with keycard, set PIN, etc) would be added in much smaller chunks of code.

### Testing notes:

PIN screen is missing (not implemented yet) so we use auto-generated PIN for now.
Also, after setup process has finished, nothing happens as login with keycard is not ready yet. 

#### Platforms
- Android only

### Steps to test:
- Open Status
- Go to Create new account
- Choose Keycard

status: ready

<blockquote><div> Figma</div><div><strong><a href="https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Hardwallet-light-flows">Figma</a></strong></div><div>Created with Figma</div></blockquote>

### CODE REVIEW

File: src/status_im/hardwallet/core.cljs [hunk 1]
```diff
@@ -28,21 +48,30 @@
...
+(fx/defn error-button-pressed [{:keys [db] :as cofx}]
+  (let [return-to-step (get-in db [:hardwallet :return-to-step] :begin)]
+    (fx/merge cofx
+              {:db (assoc-in db [:hardwallet :setup-step] return-to-step)}
+              (when-not return-to-step
+                (navigation/navigate-to-cofx :hardwallet-connect nil)))))
+
+(fx/defn load-pairing-screen [{:keys [db]}]
+  {:db       (assoc-in db [:hardwallet :setup-step] :pairing)
+   :dispatch ^:flush-dom [:hardwallet/pair]})
```
Comment (flexsurfer) [inline src/status_im/hardwallet/core.cljs]: why do you need this `^:flush-dom` ?

File: src/status_im/hardwallet/core.cljs [hunk 2]
```diff
@@ -81,14 +112,167 @@
...
+  [{:keys [db]} error]
+  (log/debug "[hardwallet] generate mnemonic error: " error)
+  {:db (-> db
+           (assoc-in [:hardwallet :setup-step] :error)
+           (assoc-in [:hardwallet :return-to-step] :card-ready)
+           (assoc-in [:hardwallet :setup-error] error))})
+
+(fx/defn on-pin-validated [{:keys [db] :as cofx}]
+  (let [pin (get-in db [:hardwallet :pin :original])]
+    (fx/merge cofx)))
```
Comment (flexsurfer) [inline src/status_im/hardwallet/core.cljs]: ?

File: src/status_im/ui/components/colors.cljs [hunk 3]
```diff
@@ -31,6 +31,7 @@
```
Comment (flexsurfer) [inline src/status_im/ui/components/colors.cljs]: are these approved by @denis-sharypin and @errorists ?

**Review by flexsurfer — APPROVED** [review]
it's too big to review :)

File: src/status_im/ui/components/colors.cljs (hunk 3 above)
Comment (errorists) [reply src/status_im/ui/components/colors.cljs]: yeah, what is this blue transparent 30, green transparent 10, are they new colours in the component lib @denis-sharypin ?

File: src/status_im/hardwallet/core.cljs [hunk 4]
```diff
@@ -51,8 +80,10 @@
 (defn- proceed-to-pin-confirmation [fx]
   (assoc-in fx [:db :hardwallet :pin :enter-step] :confirmation))
 
-(defn- pin-match [fx]
-  (assoc-in fx [:db :hardwallet :pin :status] :validating))
+(defn- pin-match [{:keys [db]}]
+  {:db                   (assoc-in db [:hardwallet :pin :status] :validating)
+   :utils/dispatch-later [{:ms       3000
+                           :dispatch [:hardwallet.callback/on-pin-validated]}]})
```
Comment (goranjovic) [inline src/status_im/hardwallet/core.cljs]: why the 3s delay?

File: src/status_im/ui/components/colors.cljs (hunk 3 above)
Comment (denis-sharypin) [reply src/status_im/ui/components/colors.cljs]: @errorists  With blue transparent 30 it is my bad, I went wild.  @dmitryn could you replace it with `blue-light`, please?

Comment (denis-sharypin) [reply src/status_im/ui/components/colors.cljs]: @flexsurfer green transparent 10, was added to our library after discussion with designers

File: src/status_im/hardwallet/core.cljs (hunk 4 above)
Comment (dmitryn) [reply src/status_im/hardwallet/core.cljs]: @goranjovic Removed this code as we don't allow set PIN for now

File: src/status_im/hardwallet/core.cljs (hunk 2 above)
Comment (dmitryn) [reply src/status_im/hardwallet/core.cljs]: @flexsurfer Removed this code as we don't allow set PIN for now

File: src/status_im/ui/components/colors.cljs (hunk 3 above)
Comment (dmitryn) [reply src/status_im/ui/components/colors.cljs]: @denis-sharypin changed to `blue-light`

File: src/status_im/hardwallet/core.cljs (hunk 1 above)
Comment (dmitryn) [reply src/status_im/hardwallet/core.cljs]: @flexsurfer it was leftover from experiments with rendering (js thread was blocked by keycard api native call, and this metadata should have allowed to view gets rendered before native call blocks it. But it didn't help, and i switch to running native calls in separate thread.

Removed this metadata , it's not needed it.

File: src/status_im/hardwallet/core.cljs (hunk 4 above)
Comment (dmitryn) [reply src/status_im/hardwallet/core.cljs]: @goranjovic i've addressed your and other guys comments, please take a look.

**Review by goranjovic — APPROVED** [review]

**Comment by asemiankevich** [comment]
used: Huawei P10 lite, 5.2" , android 8

@denis-sharypin 

1- is there anything we can do to place all the info on 1 screen? now the information does not fit the screen so i should scroll it. I feel it not very comfortable. This also happens on another screens.

image

2 - the copy on main screen seems to be incomplete and differs from figma. 


3- also the text on the very last screen is hidden under 3 dots which is weird , we have enough place to show all the placeholder here:

image


@dmitryn seems something is wrong with translations? Can't proceed when in Russian


image


also i noticed if i did not complete the process and closed the app , the NFC could hardly detect the chip even i did not move the phone and card at all.

@chadyj can we create a project in Lokalise to have this nice keycard translated too?

**Comment by chadyj** [comment]
> can we create a project in Lokalise to have this nice keycard translated too?

The strings will automatically be imported to Lokalise when this is merged :)

**Comment by denis-sharypin** [comment]
@asemiankevich Good catches!

1. That was made based on the feedback from UXR sessions. Participants wanted to see an overview of the setup steps.
2. @dmitryn Update it, please.
3. Yep, please @dmitryn make the text container wider.

**Comment by asemiankevich** [comment]
@dmitryn will it be hidden under dev mode or be visible as is in nightlies?

**Comment by dmitryn** [comment]
>@dmitryn will it be hidden under dev mode or be visible as is in nightlies?

@asemiankevich it will be hidden everywhere

**Comment by annadanchenko** [comment]
@dmitryn on my Google Pixel XL, Android 9 I'm getting "Something went wrong" error on Preparing card step. Tried 3 times already with 1 card.
Logcats of last 2 attempts:
[card2.txt](https://github.com/status-im/status-react/files/2690869/card2.txt)
[card3.txt](https://github.com/status-im/status-react/files/2690870/card3.txt)

**Comment by dmitryn** [comment]
@annadanchenko in card2.txt log i see ` StatusKeycard: Tag was lost.` that means that connection to nfc was lost. Could you try putting phone on the card during setup?

**Comment by dmitryn** [comment]
Also, we have test app https://github.com/status-im/keycard-installer-android/releases/download/2.0.0/keycard-installer-debug.apk which can show where NFC "sweet spot" is located on your device

<blockquote></blockquote>

**Pull request merged**